非同期処理(JavaScript)

Judeeeee opened the pull request:
# 概要
このPRは[非同期処理](https://bootcamp.fjord.jp/practices/204)の提出物です。

## チェック項目

Judeeeee committed:
npm installでセットアップ

Judeeeee committed:
コールバック関数のエラー有無ファイル

Judeeeee committed:
分割したファイル間で共通利用できるようにクエリを1ファイルにまとめた

Judeeeee committed:
Promiseを使ってラップした

Judeeeee committed:
Promiseで作った関数をasync/awaitの形式で使う

Judeeeee committed:
Pretterで指摘された箇所の修正

Judeeeee committed:
node-sqlite3 の関数をPromiseベースに変更。また、each()の代わりにall()を使用する。

each()だと、戻り値が配列ではないので、扱いにくい。
all()だと配列で返ってくるのでall()を採用した。

Judeeeee commented on 03.asynchronous/commonQuery.js:
```diff
@@ -1,43 +1,44 @@
-        "CREATE TABLE IF NOT EXISTS books(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)";
-      db.run(sql, function () {
-        resolve();
+  constructor() {
+    this.db = new sqlite3.Database(":test:");
+  }
+
```
node-sqlite3 の関数のように、`db.run()`の形で呼びたいので`construcor()`に記述しました。

Judeeeee commented on 03.asynchronous/commonQuery.js:
```diff
@@ -1,43 +1,44 @@
+        if (err) {
+          reject(err);
+        } else {
+          resolve({ lastID: this.lastID });
```
`run()`は、
- DBの作成
- DBの削除
- `insert`

の3つの処理で使っています。
`insert`の後は`lastID`が戻り値として欲しいのですが、DBの作成・削除では不要だと考えてこの記述にしました。

Judeeeee committed:
インメモリのデータベースを作成

ref: https://github.com/TryGhost/node-sqlite3/wiki/API#new-sqlite3databasefilename--mode--callback

Judeeeee committed:
独自の関数を定義せずに、エラーあり・なしの2つのファイルを作成した

Judeeeee committed:
エラーファイルを2つに分ける必要はない

Judeeeee committed:
クラス定義を止めて、関数を定義した

Judeeeee committed:
npm run fixで体裁を整えた

cafedomancer commented on 03.asynchronous/noErrorCallbackFunction.js:
```diff
@@ -0,0 +1,23 @@
+import sqlite3 from "sqlite3";
+const db = new sqlite3.Database(":memory:");
```
import とプログラムの間には空行を入れてください。

cafedomancer commented on 03.asynchronous/noErrorCallbackFunction.js:
```diff
@@ -0,0 +1,23 @@
+  });
+};
+
+noErrorCallbackFunction(db);
```
`db` が、関数の外で定義されていて、関数の引数として受け取るようになっている理由がよく分からないです。`noErrorCallbackFunction` が main に相当する関数なのであれば、定義の処理もその関数の中に入れるべきだと思います。そもそも main 関数自体は必要ないですが、定義したいのであればお好みで定義してもよいです。その場合は、分かりやすいように `main` という名前にしたほうがよいと思います。

cafedomancer commented on 03.asynchronous/noErrorCallbackFunction.js:
```diff
@@ -0,0 +1,23 @@
+
+const noErrorCallbackFunction = (db) => {
+  const createTableQuery =
+    "CREATE TABLE IF NOT EXISTS books(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)";
```
変数に入れる意味がないような気がしました。再利用されているわけでもないし、説明的になっているわけでもないからです。他の SQL クエリについても同様です。また、この時点でテーブルが存在していることはないので、`IF NOT EXISTS` を付けても意味がないように思います。

cafedomancer commented on 03.asynchronous/noErrorCallbackFunction.js:
```diff
@@ -0,0 +1,23 @@
+  const createTableQuery =
+    "CREATE TABLE IF NOT EXISTS books(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)";
+  db.run(createTableQuery, () => {
+    const insertQuery = "INSERT INTO books(title) VALUES('OK')";
```
`books` テーブルの `title` カラムに値を格納するのであれば、`'OK'` のような値ではなく、本のタイトルであることが分かるようなものを入れましょう。

cafedomancer commented on 03.asynchronous/noErrorCallbackFunction.js:
```diff
@@ -0,0 +1,23 @@
+  db.run(createTableQuery, () => {
+    const insertQuery = "INSERT INTO books(title) VALUES('OK')";
+    db.run(insertQuery, function () {
+      console.log(`自動採択されたID ${this.lastID}`);
```
「採択」ではなく「採番」かと。

cafedomancer commented on 03.asynchronous/noErrorCallbackFunction.js:
```diff
@@ -0,0 +1,23 @@
+    db.run(insertQuery, function () {
+      console.log(`自動採択されたID ${this.lastID}`);
+      const selectRecordsQuery = "SELECT * FROM books";
+      db.all(selectRecordsQuery, function (err, rows) {
```
関数の内部で `this` を使っていないので、function 式ではなくアロー関数式でよいと思います。また、`err` のような使っていない変数には、使っていないことが分かるような命名規則を適用してください。

cafedomancer commented on 03.asynchronous/ErrorCallbackFunction.js:
ファイル名として lowerCamelCase と UpperCamelCase が混在しています。一貫性のある命名を行ってください。

cafedomancer commented on 03.asynchronous/ErrorCallbackFunction.js:
```diff
@@ -0,0 +1,31 @@
+            db.run(dropQuery);
+          }
+        });
+      }
```
エラーがある場合に後続の処理が行われなくなっています。エラーの有無にかかわらず、以下の 4 つの処理を行えるようにしてください。

1. テーブルを作成する
1. エラーを伴うレコードの追加を行い、発生したエラーのエラーメッセージを標準エラー出力に出力する
1. エラーを伴うレコードの取得を行い、発生したエラーのエラーメッセージを標準エラー出力に出力する
1. テーブルを削除する

cafedomancer commented on 03.asynchronous/ErrorCallbackFunction.js:
```diff
@@ -0,0 +1,31 @@
+import sqlite3 from "sqlite3";
+const db = new sqlite3.Database(":memory:");
+
+const ErrorCallbackFunction = (db) => {
```
関数名には UpperCamelCase ではなく lowerCamelCase を使うのが一般的です。UpperCamelCase が使われているとクラスに見えます。

cafedomancer commented on 03.asynchronous/ErrorCallbackFunction.js:
```diff
@@ -0,0 +1,31 @@
+    const insertQuery = "INSERT INTO books(title) VALUES(NULL)";
+    db.run(insertQuery, function (err) {
+      if (err) {
+        console.error(err);
```
エラーそのものではなくエラーメッセージを表示してください。

cafedomancer commented on 03.asynchronous/commonQuery.js:
```diff
@@ -0,0 +1,38 @@
+      if (err) {
+        reject(err);
+      } else {
+        resolve({ lastID: this.lastID });
```
元の `run` 関数では `this` に属する `lastID` 以外のプロパティも結果として使えていたはずですが、この新しい関数では `lastID` しか使えなくなっています。

cafedomancer commented on 03.asynchronous/commonQuery.js:
```diff
@@ -0,0 +1,38 @@
+      resolve();
+    }
+  });
+};
```
この関数は run や all とは種類の異なる関数なので、このファイルに定義すべきではないように思います。そもそも、Promise の内部で行われている処理は非同期処理ではないので、Promise にする必要も関数にする必要もないように思います。

cafedomancer commented on 03.asynchronous/noErrorPromiseFunction.js:
```diff
@@ -0,0 +1,17 @@
+run(db, createTableQuery)
+  .then(() => run(db, insertTableQuery))
+  .then(({ lastID }) => console.log(`自動採択されたID ${lastID}`))
+  .then(() => all(db, getRecordsQuery))
```
`console.log` は非同期処理ではないので、その完了を `then` で待つ必要はないはずです。

cafedomancer commented on 03.asynchronous/noErrorPromiseFunction.js:
```diff
@@ -0,0 +1,17 @@
+  .then(() => all(db, getRecordsQuery))
+  .then((rows) => display(rows))
+  .then(() => run(db, dropTableQuery))
+  .catch((e) => console.error(e));
```
エラーは起こらない想定なので、この catch は不要だと思います。

cafedomancer commented on 03.asynchronous/ErrorPromiseFunction.js:
```diff
@@ -0,0 +1,17 @@
+  .then(() => all(db, errorGetRecordsQuery))
+  .then((rows) => display(rows))
+  .then(() => run(db, dropTableQuery))
+  .catch((e) => console.error(e));
```
エラーありのコールバック版のプログラムにもコメントした通り、1 から 4 のすべての処理を実行できるようにしてください。

cafedomancer commented on 03.asynchronous/noErrorCallbackFunction.js:
他のファイルから使われる関数が定義されているだけなら `xxxFunction.js` という名前でもまだ良いかもしれません (名前が具体に寄り過ぎていて良いとは言えない) が、定義に加えて実行まで行ってしまっているので、名前を見直したほうがよい気がしました。

cafedomancer commented on 03.asynchronous/noErrorAsyncFunctoin.js:
```diff
@@ -0,0 +1,20 @@
+import sqlite3 from "sqlite3";
+const db = new sqlite3.Database(":memory:");
+
+const noErrorAsyncFunction = async function () {
```
Top-level await を使えば async 関数の定義が不要になると思います。

cafedomancer commented on 03.asynchronous/ErrorAsyncFunctoin.js:
```diff
@@ -0,0 +1,28 @@
+  try {
+    const { lastID } = await run(db, errorInsertTableQuery);
+    console.log(`自動採択されたID ${lastID}`);
+  } catch (err) {
+    console.error(err);
+  }
```
これだとすべての例外が無条件に捕捉され、握り潰されてしまっているので行儀が良くないです。捕捉したい例外のみを捕捉するようにし、それ以外の例外は再送出するようにしてください。

cafedomancer commented on 03.asynchronous/commonQuery.js:
run や all はクエリではないので、`commonQuery.js` だと名前と実態にズレが発生していそうです。

Judeeeee committed:
ファイル名をlowerCamelCaseに統一

Judeeeee committed:
import とプログラムの間に空行を入れる

Judeeeee committed:
関数の外で定義する意味はないので、関数内で定義した

Judeeeee committed:
SQL文をここで変数に入れる意味がないので修正

Judeeeee committed:
本のタイトルを指定した

Judeeeee committed:
採番に修正

Judeeeee committed:
関数内でthisを使っていないので、アロー関数に書き換えた

Judeeeee committed:
関数名はlowerCamelCaseに統一する

Judeeeee committed:
エラーは発生しないので記述を削除

Judeeeee committed:
課題の条件通り、エラーがある場合でも後続の処理を実行させる

Judeeeee committed:
エラーそのものではなくエラーメッセージを表示させる

Judeeeee committed:
node-sqlite3のrunではchangesも返しているため、記述を追加

ref: https://github.com/TryGhost/node-sqlite3/wiki/API#runsql--param---callback

Judeeeee committed:
表示の責務を持つ関数は、SQLを実行する関数と同一のファイルにまとめない

Judeeeee committed:
console.log は非同期処理ではない

Judeeeee committed:
Top-level awaitを使う

Judeeeee committed:
捕捉したい例外のみを捕捉

ref: https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Statements/try...catch#%E6%9D%A1%E4%BB%B6%E4%BB%98%E3%81%8D_catch_%E3%83%96%E3%83%AD%E3%83%83%E3%82%AF

Judeeeee committed:
エラーありのPromise版も捕捉したくないエラーを拾っていたので修正

Judeeeee committed:
IF NOT EXISTSの対応漏れを修正

Judeeeee committed:
クエリを実行する関数群なので、databaseFunctionsに変更

Judeeeee committed:
関数の実行まで行っているので、ファイル名からfunctionを削除した

Judeeeee committed:
npm run fixを実行して体裁を整えた

cafedomancer commented on 03.asynchronous/noErrorCallback.js:
```diff
@@ -0,0 +1,21 @@
+import sqlite3 from "sqlite3";
+
+const noErrorCallbackFunction = () => {
```
コメントがないのでどう判断したのか分かりませんでしたが、main 相当の関数が必要なのだと判断したとして、関数の名前を `main` にしたほうがよいというフィードバックが反映されていないように見えます。https://github.com/Judeeeee/js-practices/pull/3#discussion_r1501334088

Judeeeee replied:
ご指摘ありがとうございます。仰る通りです。
[4737ae6](https://github.com/Judeeeee/js-practices/pull/3/commits/4737ae66aa2559e80c5cce34ed1abb717431f108)のcommitで対応しました。

cafedomancer commented on 03.asynchronous/errorCallback.js:
```diff
@@ -0,0 +1,29 @@
+            for (const row of rows) {
+              console.log(`${row.id} ${row.title}`);
+            }
+            db.run("DROP TABLE books");
```
この処理が `else` の中に含まれていると、`db.all` でエラーが無かったケースにしか実行されないです。エラーの有無にかかわらず、実行できるようにしてください。

Judeeeee replied:
[1a79315](https://github.com/Judeeeee/js-practices/pull/3/commits/1a79315025777fbd5bed7d698ad449a3e30a8c36)で対応いたしました。

cafedomancer commented on 03.asynchronous/noErrorPromise.js:
```diff
@@ -0,0 +1,24 @@
+
+const db = new sqlite3.Database(":memory:");
+
+const createTableQuery =
+  "CREATE TABLE books(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)";
+const insertTableQuery = "INSERT INTO books(title) VALUES('吾輩は猫である')";
+const getRecordsQuery = "SELECT * FROM books";
+const dropTableQuery = "DROP TABLE books";
```
以前にコメントした内容が修正されていないように見えます。https://github.com/Judeeeee/js-practices/pull/3#discussion_r1501334322

Judeeeee replied:
ご指摘ありがとうございます。[a1caf88](https://github.com/Judeeeee/js-practices/pull/3/commits/a1caf88021911998c10e99d56ba6ed8b1f6955bf)で修正いたしました。

cafedomancer commented on 03.asynchronous/noErrorPromise.js:
```diff
@@ -0,0 +1,24 @@
+const insertTableQuery = "INSERT INTO books(title) VALUES('吾輩は猫である')";
+const getRecordsQuery = "SELECT * FROM books";
+const dropTableQuery = "DROP TABLE books";
+const display = (rows) => {
+  for (const row of rows) {
+    console.log(`${row.id} ${row.title}`);
+  }
+};
```
重複した処理が共通化されているわけでもないので、この処理を関数にする必要はないと思うのですが、どうして関数にしていますか？

Judeeeee replied:
可読性のために関数にしていましたが、1回しか実行されない処理なので関数化をやめました。
[fd0f559](https://github.com/Judeeeee/js-practices/pull/3/commits/fd0f5590777f3a2e2e6abe15991c3dfe6a71cf89)で修正いたしました🙆‍♀️

cafedomancer commented on 03.asynchronous/noErrorPromise.js:
```diff
@@ -0,0 +1,24 @@
+    console.log(`自動採番されたID ${lastID}`);
+    return all(db, getRecordsQuery);
+  })
+  .then((rows) => display(rows))
+  .then(() => run(db, dropTableQuery));
```
`display` で行っている処理は非同期処理ではないので、その完了を `then` で待つ必要はないと思います。

Judeeeee replied:
[d65eb5c](https://github.com/Judeeeee/js-practices/pull/3/commits/d65eb5cabc02ace1f21b531f77d5ef4025881346)で対応したしました。

cafedomancer commented on 03.asynchronous/databaseFunctions.js:
```diff
@@ -0,0 +1,25 @@
+      if (err) {
+        reject(err);
+      } else {
+        resolve({ lastID: this.lastID, changes: this.changes });
```
`lastID` と `changes` をプロパティとして持つオブジェクトは `this` そのものだと思うのですが、`this` で使うのではなく、そのようなオブジェクトを新たに作成しているのはなぜですか？

Judeeeee replied:
`this`と記述しただけでは実行できないと思い込んでおり、新たにオブジェクトを作成していました。
[4e312ee](https://github.com/Judeeeee/js-practices/pull/3/commits/4e312ee0efa8b2336b2b3148fb87fa4b4261f89f)で対応いたしました。

cafedomancer commented on 03.asynchronous/noErrorPromise.js:
```diff
@@ -0,0 +1,24 @@
+import { run, all } from "./databaseFunctions.js";
+import sqlite3 from "sqlite3";
```
そうしたほうがいいという明確なルールがあるわけではありませんが、Node.js のライブラリ、NPM パッケージ、自分で書いたプログラムの順に import を書いておくのがよいと思います。

Judeeeee replied:
[9d3b8c1](https://github.com/Judeeeee/js-practices/pull/3/commits/9d3b8c1e33ee3942b388cc8c81297ea53611dde4)で修正いたしました。

cafedomancer commented on 03.asynchronous/errorPromise.js:
```diff
@@ -0,0 +1,44 @@
+
+run(db, createTableQuery)
+  .then(() => run(db, errorInsertTableQuery))
+  .then(
+    (lastID) => {
+      console.log(`自動採番されたID ${lastID}`);
+    },
+    (err) => {
+      console.log(err.message);
+    },
+  )
```
`then` に `onRejected` を指定するのでもよいのですが、`catch` を使ったほうが見やすいかなーと思います。

Judeeeee replied:
[7a47299](https://github.com/Judeeeee/js-practices/pull/3/commits/7a472991a0d168a00f92e4bdf9c52c588eebf132)のcommitで`catch`を使う形へ修正いたしました。

cafedomancer commented on 03.asynchronous/errorPromise.js:
```diff
@@ -0,0 +1,44 @@
+    },
+  )
+  .then(
+    () => {
+      return all(db, errorGetRecordsQuery);
+    },
```
アロー関数の本文が `return` しかないときは、ブロック文体ではなく簡潔文体を使いましょう。

Judeeeee replied:
[c20f5c0](https://github.com/Judeeeee/js-practices/pull/3/commits/c20f5c0d20b82292483e423236256fddeba03f18)で対応いたしました。

cafedomancer commented on 03.asynchronous/noErrorPromise.js:
```diff
@@ -0,0 +1,24 @@
+
+run(db, createTableQuery)
+  .then(() => run(db, insertTableQuery))
+  .then(({ lastID }) => {
```
フィードバックではなく質問なのですが、`this` を受け取るのではなく、`lastID` を分割代入で受け取っているのはなぜですか？

Judeeeee replied:
> this を受け取るのではなく、lastID を分割代入で受け取っているのはなぜですか？

当時分割代入の理解が浅くで、`{}`と`()`の使い分けがわかっていなかった、というのが理由です😓

cafedomancer commented on 03.asynchronous/noErrorAsync.js:
```diff
@@ -0,0 +1,21 @@
+const { lastID } = await run(db, insertTableQuery);
+console.log(`自動採番されたID ${lastID}`);
+const rows = await all(db, getRecordsQuery);
+await display(rows);
```
`display` は async 関数ではないので `await` を付けても意味がないです。

Judeeeee replied:
[fcadcd5](https://github.com/Judeeeee/js-practices/pull/3/commits/fcadcd53dc9c6a406eec10aefa7be2dd897b5b39)で修正いたしました。

cafedomancer commented on 03.asynchronous/errorAsync.js:
```diff
@@ -0,0 +1,35 @@
+  const { lastID } = await run(db, errorInsertTableQuery);
+  console.log(`自動採番されたID ${lastID}`);
+} catch (err) {
+  if (err instanceof Object) {
```
`Object` だと捕捉されるものが広くなりすぎてしまいませんかね...？ 以下のようなコードでも捕捉されてしまうような気がします。

```js
try {
  throw {};
} catch (err)
  ...
}
```

Judeeeee replied:
仰る通りです👀
[1acf1f8](https://github.com/Judeeeee/js-practices/pull/3/commits/1acf1f86c93ffd94fc5b3a20e074db4612a0519b)で修正いたしました。

Judeeeee committed:
関数自体が必要ないので削除

Judeeeee committed:
クエリを変数に入れない

Judeeeee committed:
不要なawaitの削除

Judeeeee committed:
エラーの有無に関わらずtableは削除したい

Judeeeee committed:
可読性のために関数にしていたが、1回しか実行されない処理なので関数化をやめた

Judeeeee committed:
完了をthenで待つ必要はない

Judeeeee committed:
importの順序を入れ替える

Judeeeee committed:
見やすさのためにonRejectedからcatchに変更

Judeeeee committed:
簡潔文体を使う

Judeeeee committed:
対象のエラーを絞る

Judeeeee committed:
オブジェクトを新たに作成する必要はない

Judeeeee commented:
レビューありがとうございました🙇
対応漏れがないように、ご指摘に対応するcommitを返信で記載させていただきました。
引き続きよろしくお願いいたします。

Judeeeee committed:
VScodeでLintとPrettierが動くようにした

cafedomancer commented on 03.asynchronous/package.json:
```diff
@@ -6,7 +6,8 @@
     "eslint-config-prettier": "^9.0.0",
-    "prettier": "^3.1.0"
+    "prettier": "^3.1.0",
+    "sqlite3": "^5.1.6"
```
見落としていましたが、`sqlite3` は `devDependencies` ではなく `dependencies` に追加すべきです。

Judeeeee replied:
ありがとうございます。
[1f767e4](https://github.com/Judeeeee/js-practices/pull/3/commits/1f767e436d610a990b3cacb2dd2a71e2bcffe2e5)で修正いたしました。

cafedomancer commented on 03.asynchronous/noErrorCallback.js:
```diff
@@ -0,0 +1,17 @@
+import sqlite3 from "sqlite3";
+
+const db = new sqlite3.Database(":memory:");
+db.run(
```
`const db = ...` のあとに空行があったりなかったりします。どちらかに揃えてスタイルを一貫させてください。

Judeeeee replied:
[a52a191](https://github.com/Judeeeee/js-practices/pull/3/commits/a52a191605d4087b7cdd6fdeeaaf7e8f43767e4b)でスタイルを一貫しました。

cafedomancer commented on 03.asynchronous/noErrorPromise.js:
```diff
@@ -0,0 +1,20 @@
+    for (const row of rows) {
+      console.log(`${row.id} ${row.title}`);
+    }
+    run(db, "DROP TABLE books");
```
現在は後続の処理がないのでこれでも問題ないですが、後続の処理を書く場合は `run` で得られる promise を `return` しておかないといけないと思うので、先にそうしておいてください。

Judeeeee replied:
[0106f22](https://github.com/Judeeeee/js-practices/pull/3/commits/0106f2209e9d3593ccc97955825e8317634a371e)で修正しました🙆‍♀️

cafedomancer commented on 03.asynchronous/errorPromise.js:
```diff
@@ -0,0 +1,26 @@
+  "CREATE TABLE books(id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)",
+)
+  .then(() => run(db, "INSERT INTO books(title) VALUES(NULL)"))
+  .then((lastID) => {
```
なにかおかしいような...。

Judeeeee replied:
[a5c2bcf](https://github.com/Judeeeee/js-practices/pull/3/commits/a5c2bcf376a90d9a9b884f3a26eeec0e35f4ac5e)で対応しました。

cafedomancer commented on 03.asynchronous/errorAsync.js:
```diff
@@ -0,0 +1,30 @@
+  const { lastID } = await run(db, "INSERT INTO books(title) VALUES(NULL)");
+  console.log(`自動採番されたID ${lastID}`);
+} catch (err) {
+  if (err.code === "SQLITE_CONSTRAINT") {
```
`err` にプロパティアクセスできないような値が入っている場合に、`err.code` がエラーになってしまうと思います。

Judeeeee replied:
[80ea206](https://github.com/Judeeeee/js-practices/pull/3/commits/80ea206ce6a5de708c026ec0e039b8c9172245b7)で修正しました。

cafedomancer commented on 03.asynchronous/errorAsync.js:
```diff
@@ -0,0 +1,30 @@
+  for (const row of rows) {
+    console.log(`${row.id} ${row.title}`);
+  }
+} catch (err) {
+  if (err.code === "SQLITE_ERROR") {
+    console.log(err.message);
+  }
+}
```
想定しない例外の場合は再送出してください。

Judeeeee replied:
[af3d369](https://github.com/Judeeeee/js-practices/pull/3/commits/af3d3693d0255ce6c307d7e564b3a3562da03899)で修正しました。

Judeeeee committed:
sqlite3をdependenciesに追加

Judeeeee committed:
空行があるスタイルに統一

Judeeeee committed:
noErrorPromiseの記述と合わせる

Judeeeee committed:
messageはerrのプロパティにあるので、codeをmessageに書き換えた

ref: https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Error/message

Judeeeee committed:
想定しないエラーの場合は再送出する

Judeeeee committed:
後続の処理を書く場合を想定してreturnを追加

cafedomancer commented on 03.asynchronous/errorAsync.js:
```diff
@@ -0,0 +1,32 @@
+  const { lastID } = await run(db, "INSERT INTO books(title) VALUES(NULL)");
+  console.log(`自動採番されたID ${lastID}`);
+} catch (err) {
+  if (err.message.includes("SQLITE_CONSTRAINT")) {
```
`err.message` はプロパティアクセスなので https://github.com/Judeeeee/js-practices/pull/3#discussion_r1522935923 であげている問題が解決していないです。

Judeeeee replied:
承知しました。
`.toString()`を使う形に変更しました。
対応commit: https://github.com/Judeeeee/js-practices/pull/3/commits/7b656030b747722c12fea6cf63b6d10c6389c7e2

cafedomancer replied:
`toString()` はメソッドで、メソッドはプロパティです。`err` に `null` や `undefined` のような値が入っているとエラーになります。

Judeeeee replied:
確かに仰る通りです🧐
[348dc9e](https://github.com/Judeeeee/js-practices/pull/3/commits/348dc9e6a47e54196fc7a73f69bc74d2035f4c9e)で修正いたしました。

Judeeeee committed:
関数を使う形に変更

Judeeeee committed:
errがnullやundefinedでないことをチェックする

ref: https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Boolean#%E8%AB%96%E7%90%86%E5%80%A4%E3%81%B8%E3%81%AE%E5%9E%8B%E5%BC%B7%E5%88%B6

cafedomancer commented on 03.asynchronous/errorAsync.js:
```diff
@@ -0,0 +1,32 @@
+  const { lastID } = await run(db, "INSERT INTO books(title) VALUES(NULL)");
+  console.log(`自動採番されたID ${lastID}`);
+} catch (err) {
+  if (err && err.toString().includes("SQLITE_CONSTRAINT")) {
```
例えば以下のようなコードでも条件を満たしてしまうと思います。

```js
const err = "SQLITE_CONSTRAINT is a constant of node-sqlite3";
err.toString().includes("SQLITE_CONSTRAINT");
```

`err` にこのような値が入っていたとして、この場合は捕捉したい例外ではないので捕捉すべきではないです。チェック方法を根本的に見直す必要があるのではと思います。今まで挙げたのはチェック方法に問題があることを示す例の一部に過ぎません。

Judeeeee replied:
修正commitは[133a051](https://github.com/Judeeeee/js-practices/pull/3/commits/133a05130e8e8377e75a08cd9d70c6828f0ec031)です。

[mdn web docの特定のエラーを処理する](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Error#%E7%89%B9%E5%AE%9A%E3%81%AE%E3%82%A8%E3%83%A9%E3%83%BC%E3%82%92%E5%87%A6%E7%90%86%E3%81%99%E3%82%8B )を参考に、

- catchされる`err`がError オブジェクトであること
- 今回補足したいエラー(`SQLITE_CONSTRAINT`と`SQLITE_ERROR`)は`Error: message`か`code`プロパティでしか特定できない
    - `Error: message`を採用しなかった理由は、`error.message.includes("SQLITE_CONSTRAINT”)`だと、エラーの文字列も含む他のエラー(SQLITE_CONSTAINT_COMMITHOOKなど)引っかかって一意に絞れないと考えたため。
    - 参考にした記事: [SQLiteエラー一覧](https://www.sqlite.org/rescode.html)

以上のことを踏まえて、`instanceof`と`code`を採用した変更になりました。

Judeeeee committed:
instanceofを使って特定のエラーを処理する

ref:
- https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Error#%E7%89%B9%E5%AE%9A%E3%81%AE%E3%82%A8%E3%83%A9%E3%83%BC%E3%82%92%E5%87%A6%E7%90%86%E3%81%99%E3%82%8B
- https://www.sqlite.org/rescode.html
